### PR TITLE
Add Animated PNG demuxer

### PR DESCRIPTION
**Pull request by bdero**
Adds a multiframe image decoder that demuxes APNG streams into regular PNG-compliant streams. The actual color decoding is done by Skia's existing PNG decoder.

A single lazy pass is performed over the full stream as the frames are decoded for the first time.

Resolves https://github.com/flutter/flutter/issues/37247.

### CODE REVIEW

**Comment by bdero** [comment]
At this point, we have pretty reasonable coverage of the spec. Both blending ops and 2/3 of the dispose ops are supported (APNG_DISPOSE_OP_PREVIOUS is the one that's missing at the moment). I think the vast majority of real world APNGs display correctly now.

https://user-images.githubusercontent.com/919017/153082672-de73c293-0d3a-40c0-ac88-9f348045c305.mov

(Tests are [this](https://philip.html5.org/tests/apng/tests.html) and [this](https://davidmz.github.io/apng-canvas/))

Known issues:
* There's a multiframecodec bug where stopped animations always advance by 1 frame when off-screen images reappear.
* APNG_DISPOSE_OP_PREVIOUS behaves like APNG_DISPOSE_OP_BACKGROUND.
* Ancillary chunks that can possibly transform frame regions are ignored -- I think these cases are somewhat pathological in practice and so I consider it out of scope for now.

**Comment by toneyzeng** [comment]
@bdero So happy to hear that you are doing for APNG, and what is the progress now?

**Comment by sveinbjornt** [comment]
What's the status on this?

**Comment by chinmaygarde** [comment]
@bdero Is this still on your plate?

**Comment by everskies** [comment]
@bdero Super excited to see this come together! Is there anything else required besides a suite of tests to get this into the engine? I'd be more than willing to add tests once you're happy with the implementation

**Comment by chinmaygarde** [comment]
Closing this as stale and there are conflicts. Please re-open if progress can be made.

**Comment by skia-gold** [comment]
Gold has detected about 247 new digest(s) on patchset 12.
View them at https://flutter-engine-gold.skia.org/cl/github/31098

**Comment by bdero** [comment]
I haven't added tests for this yet, but the implementation for this is ready for review.

**Comment by bdero** [comment]
Since the last time this was closed/reopened, Skia made the internal software rasterizing utilities private, so I had to add these in. Certain frame formats like greyscale textures are no longer supported because of that, but we can easily add those corner cases back in later.

**Comment by bdero** [comment]
Oh, and this also fixes https://github.com/flutter/flutter/issues/61150.

**Comment by eximius313** [comment]
@bdero will this PR help to adress [this feature](https://github.com/flutter/flutter/issues/37247)?

**Comment by chinmaygarde** [comment]
Ping @jason-simmons

File: lib/ui/painting/image_generator_apng.h [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    PNG_FIELD(uint32_t, height)
+    PNG_FIELD(uint32_t, x_offset)
+    PNG_FIELD(uint32_t, y_offset)
+    PNG_FIELD(uint16_t, delay_num)
+    PNG_FIELD(uint16_t, delay_den)
+    PNG_FIELD(uint8_t, dispose_op)
+    PNG_FIELD(uint8_t, blend_op)
+  };
+
+  /// @brief  The first PNG frame is always the "default" PNG frame. Absense of
```
Comment (jason-simmons) [inline lib/ui/painting/image_generator_apng.h:103]: typo: "absence"

File: lib/ui/painting/image_generator_apng.h [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+  static bool IsValidChunkHeader(const void* buffer,
+                                 size_t size,
+                                 const ChunkHeader* chunk);
+
+  static const ChunkHeader* GetNextChunk(const void* buffer,
+                                         size_t size,
+                                         const ChunkHeader* current_chunk);
+
+  /// @brief  This is a utility template for casting a png buffer pointer to a
+  ///         chunk header. It's primary purpose is to statically insert runtime
```
Comment (jason-simmons) [inline lib/ui/painting/image_generator_apng.h:148]: typo: "its"

File: lib/ui/painting/image_generator_apng.cc [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+  }
+
+  return true;
+}
+
+std::unique_ptr<ImageGenerator> APNGImageGenerator::MakeFromData(
+    sk_sp<SkData> data) {
+  // Ensure the buffer is large enough to at least contain the PNG signature
+  // and a chunk header.
+  if (data->size() < 8 + sizeof(ChunkHeader)) {
```
Comment (jason-simmons) [inline lib/ui/painting/image_generator_apng.cc:202]: Define a constant for the PNG signature size

File: lib/ui/painting/image_generator_apng.cc [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+    return nullptr;
+  }
+
+  // Demux the first image in the APNG chunk stream in order to interpret
+  // extent and blending info immediately.
+  std::optional<APNGImage> default_image;
+  const void* next_chunk_p;
+  std::tie(default_image, next_chunk_p) =
+      DemuxNextImage(data_p, data->size(), header.value(), first_chunk_p);
+  if (default_image == std::nullopt) {
```
Comment (jason-simmons) [inline lib/ui/painting/image_generator_apng.cc:251]: use `has_value()` here for consistency

File: lib/ui/painting/image_generator_apng.cc [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+  const ChunkHeader* chunk = reinterpret_cast<const ChunkHeader*>(data_p + 8);
+  if (!IsValidChunkHeader(data_p, data->size(), chunk) ||
+      chunk->get_data_length() != sizeof(ImageHeaderChunkData) ||
+      chunk->get_type() != kImageHeaderChunkType) {
+    return nullptr;
+  }
+
+  // Walk the chunks to find the "animation control" chunk. If an "image data"
+  // chunk is found first, this PNG is not animated.
+  do {
```
Comment (jason-simmons) [inline lib/ui/painting/image_generator_apng.cc:221]: Move the GetNextChunk call to the start of the loop.  The initial chunk is known to be an image header chunk and does not need to be checked.

Perhaps something like:
```
while (true) {
  chunk = GetNextChunk();
  if (chunk == nullptr)
    return nullptr;
  if (chunk->get_type() == ImageData)
    return nullptr;
  if (chunk->get_type() == AnimationControl)
    break;
}

File: lib/ui/painting/image_generator_apng.cc [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+  {
+    uint8_t* write_cursor =
+        static_cast<uint8_t*>(new_png_buffer->writable_data());
+
+    // Copy the signature/header chunks
+    memcpy(write_cursor, header.data(), header.size());
+    // If this is a frame, override the width/height in the IHDR chunk.
+    if (control_data) {
+      ChunkHeader* ihdr_header =
+          reinterpret_cast<ChunkHeader*>(write_cursor + 8);
```
Comment (jason-simmons) [inline lib/ui/painting/image_generator_apng.cc:458]: Is the 8 here the size of the PNG signature?

If so, use the constant for the signature length as mentioned above

File: lib/ui/painting/image_generator_apng.cc [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+  }
+
+  // Regardless of the byte order (RGBA vs BGRA), the blending operations are
+  // the same.
+  struct Pixel {
+    uint8_t channel[4];
+
+    uint8_t GetAlpha() { return channel[3]; }
+
+    void SetAlpha(uint8_t value) { channel[3] = value; }
```
Comment (jason-simmons) [inline lib/ui/painting/image_generator_apng.cc:130]: Remove - it looks like this is unused

**Review by jason-simmons — COMMENTED** [review]
Overall LGTM

**Comment by chinmaygarde** [comment]
Ping @bdero. Think this is really close to landing.

**Comment by bdero** [comment]
Gonna rebase this today to get it over the line.

File: lib/ui/painting/image_generator_apng.cc (hunk 6 above)
Comment (bdero) [reply lib/ui/painting/image_generator_apng.cc]: Yeah, done.

**Comment by chinmaygarde** [comment]
@jason-simmons I think you need to explicitly approve instead of adding an LGTM comment to make the presubs pass. I believe all the comments have been addressed. PTAL.

**Review by jason-simmons — APPROVED** [review]

File: lib/ui/painting/image_generator.h [hunk 8]
```diff
@@ -80,7 +84,7 @@ class ImageGenerator {
   /// @return     Information about the given frame. If the image is
   ///             single-frame, a default result is returned.
   /// @see        `GetFrameCount`
-  virtual const FrameInfo GetFrameInfo(unsigned int frame_index) const = 0;
+  virtual const FrameInfo GetFrameInfo(unsigned int frame_index) = 0;
```
Comment (toneyzeng) [inline lib/ui/painting/image_generator.h:87]: Why remove the |const| modifier?

**Comment by chinmaygarde** [comment]
@bdero, this looks good to go. Can we land this? @toneyzeng has an open question though.

File: lib/ui/painting/image_generator.h (hunk 8 above)
Comment (bdero) [reply lib/ui/painting/image_generator.h:87]: `GetFrameInfo` requires decoding the image header as per the registry protocol, which is expected to have substantial side effects per the protocol.

File: lib/ui/painting/multi_frame_codec.cc [hunk 9]
```diff
@@ -107,24 +107,30 @@ sk_sp<DlImage> MultiFrameCodec::State::GetNextFrameImage(
...
-                     << " is not cached. Using " << lastRequiredFrameIndex_
-                     << " instead";
-    }
-
-    if (lastRequiredFrame_->getPixels() &&
-        CopyToBitmap(&bitmap, lastRequiredFrame_->colorType(),
-                     *lastRequiredFrame_)) {
-      prior_frame_index = requiredFrameIndex;
+                     << " is not cached. Using blank slate instead.";
+    } else {
```
Comment (dnfield) [inline lib/ui/painting/multi_frame_codec.cc:121]: I think this is a bug. Will follow up with some more details shortly.

Comment (dnfield) [reply lib/ui/painting/multi_frame_codec.cc:121]: For GIFs at least, this isn't safe. I vaguely remember discussing this before a while back. Something like the current frame can end up depending on the last frame and the framework doesn't draw both the last and current frame so it needs a complete frame.

Comment (dnfield) [reply lib/ui/painting/multi_frame_codec.cc:121]: (IOW, the GIF can require a last frame that was several frames ago but still needs to be painted)

Comment (bdero) [reply lib/ui/painting/multi_frame_codec.cc:121]: Is the SkBitmap allocation not getting initialized with zeroes? This logic change should make the copying the action of copying the last frame more restrictive, not less. We were copying the previous frame's contents in cases where we're supposed to composite atop a blank frame, which was visible in this gif as well: https://github.com/flutter/flutter/issues/61150#issuecomment-679055858

**Pull request merged**